### PR TITLE
Feature/issue #653

### PR DESCRIPTION
**Pull request by iurii-i-popov**
Late document materialization for view node with sort variables.

Task arangodb/backlog#653.

Condition: there are exist ENUMERATE_IRESEARCH_VIEW, SORT and LIMIT nodes.
If view node has primary sort then we can check all variables to be extracted from sort keys.

Every reference to a document replaces by a new variable. Materialize node will materialize the document later.

**PR contains primary sort optimization only.**

### CODE REVIEW

**Comment by iurii-i-popov** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/7159/

**Comment by iurii-i-popov** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/7225/

**Comment by iurii-i-popov** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/7229/

**Comment by iurii-i-popov** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/7230/

**Comment by iurii-i-popov** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/7231/

File: arangod/Aql/IResearchViewNode.h [hunk 1]
```diff
@@ -199,6 +237,11 @@ class IResearchViewNode final : public arangodb::aql::ExecutionNode {
   /// @brief output variable to write only non-materialized collection ids
   aql::Variable const* _outNonMaterializedColPtr;
 
+  /// @brief output variables to non-materialized document view sort references
+  ViewValuesVars _outNonMaterializedViewVars;
+
+  /// @brief calculation node with ast nodes that can be replaced by view values (e.g. primary sort)
+  std::unordered_map<aql::CalculationNode*, ViewVarsToBeReplaced> _nodesToChange;
```
Comment (gnusi) [inline arangod/Aql/IResearchViewNode.h]: let's extract this `_nodesToChange` as a separate entity, e.g.
```
class IResearchViewNode {
  ...
  struct OptimizationState {
    /// @brief calculation node with ast nodes that can be replaced by view values (e.g. primary sort)
    std::unordered_map<aql::CalculationNode*, ViewVarsToBeReplaced> _nodesToChange;

    // other related methods
  };

  OptimizationState _optState;

  OptimizationState& state() noexcept { return _optState; }
};
```

File: arangod/Aql/IResearchViewOptimizerRules.cpp [hunk 2]
```diff
@@ -263,35 +263,88 @@ bool optimizeSort(IResearchViewNode& viewNode, ExecutionPlan* plan) {
...
+bool attributesMatch(IResearchViewSort const& primarySort, latematerialized::NodeWithAttrs& node) {
+  // check all node attributes to be in sort
+  for (auto& nodeAttr : node.attrs) {
+    size_t sortFieldNum = 0;
+    for (auto const& field : primarySort.fields()) {
+      if (arangodb::basics::AttributeName::isIdentical(nodeAttr.attr, field, false)) {
+        nodeAttr.afData.number = sortFieldNum;
+        nodeAttr.afData.field = &field;
+        std::vector<arangodb::basics::AttributeName> empty;
+        nodeAttr.attr.swap(empty); // we do not need later
```
Comment (gnusi) [inline arangod/Aql/IResearchViewOptimizerRules.cpp]: why not just `nodeAttr.attr.clear()`?

File: arangod/Aql/LateMaterializedOptimizerRulesCommon.cpp [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+#include "Aql/Ast.h"
+#include "Aql/LateMaterializedOptimizerRulesCommon.h"
+
+using namespace arangodb::aql;
+
+namespace {
+
+// traverse the AST, using previsitor
+void traverseReadOnly(AstNode* node, AstNode* parentNode, size_t childNumber,
```
Comment (gnusi) [inline arangod/Aql/LateMaterializedOptimizerRulesCommon.cpp]: you can just use `Ast::traverseReadOnly(AstNode* node, std::function<...> const& preVisitor, std::function<...> const& postVisitor);`

File: arangod/Aql/IResearchViewOptimizerRules.cpp (hunk 2 above)
Comment (iurii-i-popov) [reply arangod/Aql/IResearchViewOptimizerRules.cpp]: Clear() leaves the capacity() of the vector unchanged.

File: arangod/Aql/IResearchViewNode.h [hunk 4]
```diff
@@ -171,11 +176,44 @@ class IResearchViewNode final : public arangodb::aql::ExecutionNode {
...
+  using ViewVarsToBeReplaced = std::vector<aql::latematerialized::AstAndFieldData>;
+
+  void replaceViewVariables(std::vector<aql::CalculationNode*>const& calcNodes);
+
+  void saveCalcNodesForViewVariables(std::vector<aql::latematerialized::NodeWithAttrs> const& nodesToChange);
+
+  bool canVariablesBeReplaced(aql::CalculationNode* calclulationNode) const;
+
+  void clearViewVariables() {
+    if (!_nodesToChange.empty()) {
```
[the remark this answers is not in the document]
Comment (iurii-i-popov) [reply arangod/Aql/IResearchViewNode.h]: The C++ standard does not require clear to release the memory.

File: tests/js/server/aql/aql-optimizer-rule-late-document-materialization-arangosearch-2.js [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+      let query = "FOR d IN " + svn + " SORT NOOPT(d.str) SORT d.value DESC LIMIT 10 RETURN d";
+      let plan = AQL_EXPLAIN(query).plan;
+      assertEqual(-1, plan.rules.indexOf(ruleName));
+    },
+    testNotAppliedDueToNoLimit() {
+      let query = "FOR d IN " + svn + " SORT d.value DESC RETURN d";
+      let plan = AQL_EXPLAIN(query).plan;
+      assertEqual(-1, plan.rules.indexOf(ruleName));
+    },
+    testQueryResultsWithFilter() {
```
Comment (Dronplane) [inline tests/js/server/aql/aql-optimizer-rule-late-document-materialization-arangosearch-2.js]: Let`s also add test for late materialization NOT applied if variable is used in subquery (as we added such execution branch in optimizer rule for views).

File: arangod/Aql/IResearchViewExecutor.h [hunk 6]
```diff
@@ -164,11 +167,11 @@ class IResearchViewExecutorBase {
     OutputAqlItemRow& outputRow;
     IndexIterator::DocumentCallback const callback;
 
-    inline aql::RegisterId getNmColPtrOutReg() const noexcept {
+    aql::RegisterId getNmColPtrOutReg() const noexcept {
```
[the remark this answers is not in the document]
Comment (iurii-i-popov) [reply arangod/Aql/IResearchViewExecutor.h]: A function defined entirely inside a class/struct/union definition, whether it's a member function or a non-member friend function, is implicitly an inline function.

File: arangod/Aql/IResearchViewExecutor.h [hunk 7]
```diff
@@ -164,11 +167,11 @@ class IResearchViewExecutorBase {
...
       return docOutReg;
     }
 
-    inline aql::RegisterId getNmDocIdOutReg() const noexcept {
+    aql::RegisterId getNmDocIdOutReg() const noexcept {
```
[the remark this answers is not in the document]
Comment (iurii-i-popov) [reply arangod/Aql/IResearchViewExecutor.h]: A function defined entirely inside a class/struct/union definition, whether it's a member function or a non-member friend function, is implicitly an inline function.

File: arangod/Aql/IResearchViewOptimizerRules.cpp (hunk 2 above)
Comment (iurii-i-popov) [reply arangod/Aql/IResearchViewOptimizerRules.cpp]: OK: clear() + shrink_to_fit()

File: tests/js/server/aql/aql-optimizer-rule-late-document-materialization-arangosearch-2.js (hunk 5 above)
Comment (iurii-i-popov) [reply tests/js/server/aql/aql-optimizer-rule-late-document-materialization-arangosearch-2.js]: Done

File: arangod/Aql/LateMaterializedOptimizerRulesCommon.cpp (hunk 3 above)
Comment (iurii-i-popov) [reply arangod/Aql/LateMaterializedOptimizerRulesCommon.cpp]: I need parentNode and childNumber. Ast::traverseReadOnly does not provide this information.

File: arangod/Aql/IResearchViewNode.cpp [hunk 8]
```diff
@@ -1238,41 +1388,60 @@ std::unique_ptr<aql::ExecutionBlock> IResearchViewNode::createBlock(
...
+                                                std::move(outNonMaterializedViewRegs)};
 
-      return std::make_unique<aql::ExecutionBlockImpl<aql::IResearchViewMergeExecutor<true, true>>>(
-        &engine, this, std::move(executorInfos));
-    } else {
-      if (!ordered) {
-        return std::make_unique<aql::ExecutionBlockImpl<aql::IResearchViewMergeExecutor<false, false>>>(
-          &engine, this, std::move(executorInfos));
+  TRI_ASSERT(_sort.first == nullptr || !_sort.first->empty()); // guaranteed by optimizer rule
+  TRI_ASSERT(_sort.first == nullptr || materializeType != MaterializeType::LateMaterializedWithVars);
```
[the remark this answers is not in the document]
Comment (iurii-i-popov) [reply arangod/Aql/IResearchViewNode.cpp]: Moved into getExecutorIndex().

File: arangod/Aql/IResearchViewNode.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (iurii-i-popov) [reply arangod/Aql/IResearchViewNode.h]: Changed to clear().

**Comment by iurii-i-popov** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/7269/

**Comment by iurii-i-popov** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/7296/

**Pull request merged**